cmd/disk-cache.go

```diff
@@ -0,0 +1,674 @@
+		return false
+	}
+	usedPercent := (di.Total - di.Free) * 100 / di.Total
+	return int(usedPercent) < minUsage
+}
+
+// Retusn if the disk usage is high.
+// Disk usage is high if disk used is > maxUsage
+func (c diskCache) diskUsageHigh() bool {
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
+		return true
+	}
+	usedPercent := (di.Total - di.Free) * 100 / di.Total
+	return int(usedPercent) > c.maxUsage
+}
+
+// Purge cache entries that were not accessed.
+func (c diskCache) purge() {
+	expiry := c.createTime
+
+	for {
+		for {
+			if c.diskUsageLow() {
+				break
+			}
+			d := time.Now().UTC().Sub(expiry)
+			d = d / 2
+			expiry = expiry.Add(d)
+
+			// Delete all the entries that haven't been accessed since "expiry".
+			c.db.Update(func(tx *bolt.Tx) error {
+				b := tx.Bucket([]byte(diskCacheBoltdbBucket))
+				cursor := b.Cursor()
+				for k, v := cursor.First(); k != nil; k, v = cursor.Next() {
+					entry := make(diskCacheBoltdbEntry)
+					err := json.Unmarshal(v, &entry)
+					if err != nil {
+						errorIf(err, "Unable to unmarshall")
+						continue
+					}
+					atime, err := entry.getAtime()
+					if err != nil {
+						errorIf(err, "Unable to get atime for %s", string(k))
+					}
+					if atime.After(expiry) {
+						continue
+					}
+					s := strings.SplitN(string(k), slashSeparator, 2)
+					bucket := s[0]
+					object := s[1]
+					if bucket == "" {
+						errorIf(errUnexpected, `"bucket" is empty`)
+					}
+					if object == "" {
+						errorIf(errUnexpected, `"object" is empty`)
+					}
+					if err = os.Remove(c.encodedPath(bucket, object)); err != nil {
+						errorIf(err, "Unable to remove %s", string(k))
+						continue
+					}
+					if err = os.Remove(c.encodedMetaPath(bucket, object)); err != nil {
+						errorIf(err, "Unable to remove meta file for %s", string(k))
+						continue
+					}
+					err = b.Delete(k)
+					errorIf(err, "Unable to delete %s", string(k))
+				}
+				return nil
+			})
+		}
+		<-c.purgeChan
+	}
+}
+
+// encode (bucket,object) to a hash value which is used as the file name of the cached object.
+func (c diskCache) encodedPath(bucket, object string) string {
+	return path.Join(c.dataDir, fmt.Sprintf("%x", sha256.Sum256([]byte(path.Join(bucket, object)))))
+}
+
+// encode (bucket,object) to a hash value which is used as the file name of the cached object's metadata.
+func (c diskCache) encodedMetaPath(bucket, object string) string {
+	return path.Join(c.dataDir, fmt.Sprintf("%x.json", sha256.Sum256([]byte(path.Join(bucket, object)))))
+}
+
+// Commit the cached object - rename from tmp directory to data directory.
+func (c diskCache) Commit(f *os.File, objInfo ObjectInfo, anon bool) error {
+	bucket := objInfo.Bucket
+	object := objInfo.Name
+
+	objectLock := globalNSMutex.NewNSLock(diskCacheLockingPrefix+bucket, object)
+	objectLock.Lock()
+	defer objectLock.Unlock()
+
+	encPath := c.encodedPath(bucket, object)
+	if err := os.Rename(f.Name(), encPath); err != nil {
```

Comment:
This won't work on windows.. renaming an active fd. You should use the `lock.LockedOpenFile()` which is allowed to be renamed in this manner.